### PR TITLE
Improve Tyger scalability

### PR DESCRIPTION
**Pull request by johnstairs**
This PR changes the way we interact with the Kubernetes API in order to handle many concurrent runs.

1. New runs are written to the database and we maintain an upper limit on the number runs that have Kubernetes pods created on the cluster. Once there is space, new pods are created.
2. We now make fewer calls to the API, instead using a single watch to maintain an in-memory cache of all objects.
3. Calls to show/list/watch runs are served from the database.

We are now able to handle creating and executing runs at a sustained 60 runs per second if the cluster and database are sufficiently scaled as the following video shows:

Screen Recording 2024-10-29 at 9 39 16 PM

The database migration required for this was done in a way that requires offline migration.

### CODE REVIEW

File: server/ControlPlane/Database/Migrations/LoggerExtensions.cs [hunk 1]
```diff
@@ -5,42 +5,45 @@ namespace Tyger.ControlPlane.Database.Migrations;
...
 
-    [LoggerMessage(12, LogLevel.Warning, "Newer database versions exist")]
+    [LoggerMessage(LogLevel.Warning, "Newer database versions exist")]
     public static partial void NewerDatabaseVersionsExist(this ILogger logger);
 
-    [LoggerMessage(13, LogLevel.Information, "Using most recent database version")]
+    [LoggerMessage(LogLevel.Information, "Using most recent database version")]
     public static partial void UsingMostRecentDatabaseVersion(this ILogger logger);
+
+    [LoggerMessage(LogLevel.Error, "This version of Tyger requires the database to be migrated to at least version {requiredVersion} and the server must be take onffline for this migration. The current version is {currentVersion}.")]
```
Comment (naegelejd) [inline server/ControlPlane/Database/Migrations/LoggerExtensions.cs:47]: Nit: "offline" is misspelled in this LoggerMessage (currently "onffline")

File: server/ControlPlane/Logging/LoggerExtensions.cs [hunk 2]
```diff
@@ -5,15 +5,15 @@ namespace Tyger.ControlPlane.Logging;
...
-    [LoggerMessage(1, LogLevel.Information, "Retrieving archived logs for run {run}")]
+    [LoggerMessage(LogLevel.Information, "Retrieving archived logs for run {run}")]
     public static partial void RetrievingArchivedLogsForRun(this ILogger logger, long run);
 
-    [LoggerMessage(2, LogLevel.Information, "Resuming logs after exception.")]
+    [LoggerMessage(LogLevel.Information, "Resuming logs after exception.")]
     public static partial void ResumingLogsAfterException(this ILogger logger, Exception exception);
 
-    [LoggerMessage(3, LogLevel.Warning, "Local log file is does not have expected name '{name}'.")]
+    [LoggerMessage(LogLevel.Warning, "Local log file is does not have expected name '{name}'.")]
```
Comment (naegelejd) [inline server/ControlPlane/Logging/LoggerExtensions.cs:17]: Nit: remove "is" from LoggerMessage (currently "log file *is does* not have...")

**Review by naegelejd — APPROVED** [review]
Run management looks solid. Handling of the offline db migration worked well too.

There are handful of hardcoded values, **all of which** I think are reasonable, but wanted to point them out in case any of them should be configurable:

1. `RunIndexPruner.cs`, delay period is fixed at 5 minutes.
2. `Repository.cs`, `MaxActiveRuns` hardcoded to 2000.
3. `Repository.cs`, `ListenForNewRuns` uses a hardcoded AdvisoryLockId. It is *nearly* impossible to ever have a conflict with `hashtext(idempotencyKey)`, so this is fine.
4. `DockerRunSweeper.cs`, delay period is fixed at 30 seconds.
5. `Kubernetes/RunFinalizer.cs`, delay after failing to `FinalizeRun` is fixed at 5 seconds.
6. `Kubernetes/RunStateObserver.cs`, `MaxDegreeOfParallelism` is fixed at 10 when updating Runs from observed state.

**Review by hansenms — APPROVED** [review]

**Comment by johnstairs** [comment]
> 3. `Repository.cs`, `ListenForNewRuns` uses a hardcoded AdvisoryLockId. It is _nearly_ impossible to ever have a conflict 
 @naegelejd Good catch! I will change this to the overload that takes two integers, so that there is no possible conflict.

**Comment by johnstairs** [comment]
Regarding the constants used across the code, they could be made configurable, but I think we will hold off on doing that for now.

**Pull request merged**